File: python/thorn_django/extract.py
"""Extract Django model graph and run dynamic validation checks.

Usage:
    python -m thorn_django                                      # writes .thorn/graph.json
    python -m thorn_django --settings myproject.settings         # specify settings
    python -m thorn_django --stdout                              # output to stdout
    docker compose exec app python -m thorn_django               # in Docker
"""
import argparse
import json
import logging
import os
import sys


def main():
    parser = argparse.ArgumentParser(description="Extract Django model graph for Thorn")
    parser.add_argument("--settings", default=os.environ.get("DJANGO_SETTINGS_MODULE"),
                        help="Django settings module")
    parser.add_argument("--stdout", action="store_true",
                        help="Write JSON to stdout instead of .thorn/graph.json")
    args = parser.parse_args()

    if args.settings:
        os.environ["DJANGO_SETTINGS_MODULE"] = args.settings
    if not os.environ.get("DJANGO_SETTINGS_MODULE"):
        print("Error: DJANGO_SETTINGS_MODULE not set.", file=sys.stderr)
        sys.exit(1)

    logging.disable(logging.WARNING)
    import django
    django.setup()
    logging.disable(logging.NOTSET)

    graph = extract_graph()
    diagnostics = run_dynamic_checks()
    output = {"graph": graph, "diagnostics": diagnostics}

    if args.stdout:
        json.dump(output, sys.stdout)
    else:
        thorn_dir = os.path.join(os.getcwd(), ".thorn")
        os.makedirs(thorn_dir, exist_ok=True)
        graph_path = os.path.join(thorn_dir, "graph.json")
        with open(graph_path, "w") as f:
            json.dump(output, f)
        print(f"Wrote {len(graph['models'])} models + {len(diagnostics)} diagnostics to {graph_path}", file=sys.stderr)


def extract_graph():
    from django.apps import apps
    from django.conf import settings
    models = [_extract_model(m) for m in apps.get_models()]
    return {
        "models": models,
        "installed_apps": list(apps.app_configs.keys()),
        "settings": {
            "auth_user_model": getattr(settings, "AUTH_USER_MODEL", "auth.User"),
            "databases": list(getattr(settings, "DATABASES", {}).keys()),
            "middleware": list(getattr(settings, "MIDDLEWARE", [])),
            "extra": {},
        },
    }


def _extract_model(model):
    import inspect
    meta = model._meta
    try:
        source_file = os.path.relpath(inspect.getfile(model))
    except (TypeError, OSError):
        source_file = ""
    return {
        "app_label": meta.app_label, "name": model.__name__,
        "db_table": meta.db_table, "module": model.__module__,
        "source_file": source_file, "abstract_model": meta.abstract,
        "proxy": meta.proxy,
        "fields": _extract_fields(meta),
        "relations": _extract_relations(meta),
        "managers": _extract_managers(meta),
        "parents": [p.__name__ for p in meta.parents],
        "methods": [n for n in dir(model) if not n.startswith("_") and callable(getattr(model, n, None))],
    }


def _extract_fields(meta):
    fields = []
    for f in meta.local_concrete_fields:
        fc = f.__class__
        fclass = f"{fc.__module__}.{fc.__name__}"
        if any(x in fclass for x in ("ForeignKey", "OneToOne", "ManyToMany", "Rel")):
            continue
        from django.db.models.fields import NOT_PROVIDED
        default = None
        if f.default is not NOT_PROVIDED:
            default = f"<callable: {getattr(f.default, '__name__', '?')}>" if callable(f.default) else str(f.default)
        fields.append({
            "name": f.name, "column": f.column, "field_class": fclass,
            "native_type": _field_type(fclass), "nullable": f.null, "blank": f.blank,
            "default": default, "max_length": getattr(f, "max_length", None),
            "choices": [(str(c[0]), str(c[1])) for c in (f.choices or [])],
            "validators": [f"{v.__class__.__module__}.{v.__class__.__name__}" for v in f.validators],
            "primary_key": f.primary_key, "unique": f.unique, "db_index": f.db_index,
        })
    return fields


def _extract_relations(meta):
    relations = []
    for f in meta.get_fields(include_hidden=True):
        fc = f.__class__
        fclass = f"{fc.__module__}.{fc.__name__}"
        kind = None
        for k, v in {"ForeignKey": "ForeignKey", "OneToOneField": "OneToOne", "ManyToManyField": "ManyToMany", "ManyToOneRel": "Reverse", "OneToOneRel": "ReverseOneToOne", "ManyToManyRel": "ManyToMany"}.items():
            if k in fclass: kind = v; break
        if not kind: continue
        rm = getattr(f, "related_model", None)
        if rm is None: continue
        on_delete = None
        rf = getattr(f, "remote_field", None)
        if rf and hasattr(rf, "on_delete") and rf.on_delete:
            on_delete = rf.on_delete.__name__
        relations.append({
            "name": getattr(f, "name", "") or "", "kind": kind,
            "to_model": rm.__name__, "to_model_app": rm._meta.app_label,
            "related_name": getattr(f, "related_name", "") or "",
            "related_query_name": "", "on_delete": on_delete,
            "nullable": getattr(f, "null", False), "through_model": None,
        })
    return relations


def _extract_managers(meta):
    return [{
        "name": m.name,
        "manager_class": f"{m.__class__.__module__}.{m.__class__.__name__}",
        "queryset_class": f"{m._queryset_class.__module__}.{m._queryset_class.__name__}",
        "is_default": m == meta.default_manager,
        "custom_methods": [n for n in dir(m) if not n.startswith("_") and callable(getattr(m, n, None))],
    } for m in meta.managers]


def _field_type(fc):
    for k, v in {"IntegerField": "int", "AutoField": "int", "FloatField": "float", "DecimalField": "float", "BooleanField": "bool", "DateTimeField": "datetime", "DateField": "date", "TimeField": "time", "UUIDField": "UUID", "JSONField": "Any", "BinaryField": "bytes"}.items():
        if k in fc: return v
    return "str"


# ── Dynamic validation checks ─────────────────────────────────────────────

def run_dynamic_checks():
    d = []
    d.extend(_check_django_system_checks())
    d.extend(_check_models())
    d.extend(_check_serializers())
    d.extend(_check_migrations())
    d.extend(_check_missing_migrations())
    d.extend(_check_templates())
    d.extend(_check_dotted_path_settings())
    d.extend(_check_forms())
    return d


def _check_django_system_checks():
    from django.core import checks
    diags = []
    for c in checks.run_checks(include_deployment_checks=True):
        if c.level < checks.WARNING: continue
        code = {checks.WARNING: "DV-WARN", checks.ERROR: "DV-ERR", checks.CRITICAL: "DV-CRIT"}.get(c.level, "DV-WARN")
        diags.append({"code": code, "message": f"[{c.id}] {c.msg}", "range": None, "filename": "django.checks", "line": None, "col": None})
    return diags


def _is_third_party(path_or_module):
    """Check if a path or module name belongs to a third-party package."""
    s = str(path_or_module)
    return 'site-packages' in s or 'dist-packages' in s or '/venv/' in s or '/.venv/' in s


def _check_models():
    import inspect
    from django.apps import apps
    diags = []
    for model in apps.get_models():
        meta = model._meta
        if meta.abstract: continue
        name = f"{meta.app_label}.{model.__name__}"
        try: sf, sl = os.path.relpath(inspect.getfile(model)), inspect.getsourcelines(model)[1]
        except: sf, sl = model.__module__, None
        # Skip third-party models — developers can't fix these
        if _is_third_party(sf): continue
        has_str = any("__str__" in cls.__dict__ for cls in model.__mro__ if cls.__name__ not in ("Model", "object"))
        if not has_str:
            diags.append({"code": "DV001", "message": f"Model '{name}' has no __str__.", "range": None, "filename": sf, "line": sl, "col": None})
    return diags


def _check_serializers():
    try: from rest_framework.serializers import ModelSerializer
    except ImportError: return []
    diags = []
    for cls in _all_subclasses(ModelSerializer):
        meta = getattr(cls, "Meta", None)
        if not meta: continue
        model = getattr(meta, "model", None)
        if not model: continue
        fields = getattr(meta, "fields", None)
        if not fields or fields == "__all__": continue
        if not isinstance(fields, (list, tuple)): continue
        declared = getattr(cls, "_declared_fields", {})
        for fn in fields:
            try: model._meta.get_field(fn); continue
            except: pass
            if fn in declared: continue
            if hasattr(model, fn): continue
            diags.append({"code": "DV101", "message": f"Serializer '{cls.__name__}' field '{fn}' doesn't exist on '{model.__name__}'.", "range": None, "filename": cls.__module__, "line": None, "col": None})
    return diags


def _check_migrations():
    from django.db import connection
    from django.db.migrations.loader import MigrationLoader
    diags = []
    try:
        loader = MigrationLoader(connection, ignore_no_migrations=True)
        for app, migs in loader.detect_conflicts().items():
            diags.append({"code": "DV201", "message": f"Migration conflict in '{app}': {', '.join(migs)}", "range": None, "filename": "migrations", "line": None, "col": None})
    except Exception: pass
    return diags


def _check_missing_migrations():
    """DV202: Check if model changes exist without a corresponding migration."""
    from io import StringIO
    from django.core.management import call_command
    diags = []
    try:
        out = StringIO()
        call_command('makemigrations', '--check', '--dry-run', stdout=out, stderr=out)
    except SystemExit as e:
        if e.code != 0:
            output = out.getvalue().strip()
            diags.append({
                "code": "DV202", "range": None, "line": None, "col": None,
                "message": f"Missing migrations detected: {output}",
                "filename": "migrations",
            })
    except Exception:
        pass
    return diags


def _check_templates():
    """DV401: Compile all templates to catch TemplateSyntaxError."""
    from django.template import TemplateSyntaxError
    from django.template.loader import get_template
    from django.apps import apps
    diags = []
    template_dirs = []
    try:
        from django.conf import settings
        for engine_config in settings.TEMPLATES:
            template_dirs.extend(engine_config.get('DIRS', []))
        for app_config in apps.get_app_configs():
            app_template_dir = os.path.join(app_config.path, 'templates')
            if os.path.isdir(app_template_dir):
                template_dirs.append(app_template_dir)
    except Exception:
        return diags

    for tdir in template_dirs:
        if not os.path.isdir(tdir):
            continue
        # Skip third-party template directories
        if _is_third_party(tdir):
            continue
        for root, _, files in os.walk(tdir):
            for fname in files:
                if not fname.endswith(('.html', '.txt', '.xml', '.email')):
                    continue
                full_path = os.path.join(root, fname)
                rel_path = os.path.relpath(full_path, tdir)
                try:
                    get_template(rel_path)
                except TemplateSyntaxError as e:
                    diags.append({
                        "code": "DV401", "range": None, "line": None, "col": None,
                        "message": f"Template '{rel_path}' has a syntax error: {e}",
                        "filename": full_path,
                    })
                except Exception:
                    pass  # Template not found via loader, skip
    return diags


def _check_dotted_path_settings():
    """DV601: Verify dotted-path settings are importable."""
    from django.conf import settings
    from django.utils.module_loading import import_string
    diags = []

    # Settings that contain dotted import paths
    path_settings = {
        'MIDDLEWARE': getattr(settings, 'MIDDLEWARE', []),
        'AUTHENTICATION_BACKENDS': getattr(settings, 'AUTHENTICATION_BACKENDS', []),
        'PASSWORD_HASHERS': getattr(settings, 'PASSWORD_HASHERS', []),
    }

    # DRF settings
    drf_settings = getattr(settings, 'REST_FRAMEWORK', {})
    for key in ['DEFAULT_AUTHENTICATION_CLASSES', 'DEFAULT_PERMISSION_CLASSES',
                'DEFAULT_RENDERER_CLASSES', 'DEFAULT_THROTTLE_CLASSES',
                'DEFAULT_PAGINATION_CLASS', 'DEFAULT_FILTER_BACKENDS']:
        val = drf_settings.get(key)
        if val:
            if isinstance(val, str):
                path_settings[f'REST_FRAMEWORK.{key}'] = [val]
            elif isinstance(val, (list, tuple)):
                path_settings[f'REST_FRAMEWORK.{key}'] = list(val)

    for setting_name, paths in path_settings.items():
        for path in paths:
            if not isinstance(path, str) or '.' not in path:
                continue
            try:
                import_string(path)
            except ImportError as e:
                diags.append({
                    "code": "DV601", "range": None, "line": None, "col": None,
                    "message": f"Setting {setting_name} references '{path}' which cannot be imported: {e}",
                    "filename": "settings",
                })
    return diags


def _check_forms():
    """DV501: Validate ModelForm field definitions against models."""
    try:
        from django.forms import ModelForm
    except ImportError:
        return []
    diags = []
    for cls in _all_subclasses(ModelForm):
        # Skip third-party forms
        if _is_third_party(getattr(cls, '__module__', '')):
            continue
        meta = getattr(cls, 'Meta', None)
        if not meta:
            continue
        model = getattr(meta, 'model', None)
        if not model:
            continue
        fields = getattr(meta, 'fields', None)
        if not fields or fields == '__all__':
            continue
        if not isinstance(fields, (list, tuple)):
            continue
        for fn in fields:
            try:
                model._meta.get_field(fn)
            except Exception:
                if not hasattr(model, fn):
                    diags.append({
                        "code": "DV501", "range": None, "line": None, "col": None,
                        "message": f"ModelForm '{cls.__name__}' references field '{fn}' not on model '{model.__name__}'.",
                        "filename": cls.__module__,
                    })
    return diags


def _all_subclasses(cls, seen=None):
    if seen is None: seen = set()
    result = []
    for sc in cls.__subclasses__():
        if id(sc) in seen or sc.__name__.startswith("_"): continue
        seen.add(id(sc))
        meta = getattr(sc, "Meta", None)
        if meta and getattr(meta, "abstract", False): continue
        result.append(sc)
        result.extend(_all_subclasses(sc, seen))
    return result

File: src/bridge/extract.rs
use pyo3::prelude::*;
use pyo3::types::PyDict;
use thorn_api::graph::*;

/// Extract the full model graph from a booted Django process.
pub fn extract_graph(py: Python<'_>) -> PyResult<ModelGraph> {
    let apps = py.import("django.apps")?.getattr("apps")?;

    let models_list = apps.call_method0("get_models")?;
    let mut models = Vec::new();

    for model_cls in models_list.try_iter()? {
        let model_cls = model_cls?;
        match extract_model(py, &model_cls) {
            Ok(model) => models.push(model),
            Err(e) => {
                eprintln!("Warning: failed to extract model: {e}");
            }
        }
    }

    let installed_apps = extract_installed_apps(py)?;
    let settings = extract_settings(py)?;

    Ok(ModelGraph {
        models,
        installed_apps,
        settings,
    })
}

fn extract_model(py: Python<'_>, model_cls: &Bound<'_, PyAny>) -> PyResult<DjangoModel> {
    let meta = model_cls.getattr("_meta")?;

    let app_label: String = meta.getattr("app_label")?.extract()?;
    let model_name: String = meta.getattr("model_name")?.extract()?;
    let name = capitalize(&model_name);
    let db_table: String = meta.getattr("db_table")?.extract()?;
    let module: String = model_cls.getattr("__module__")?.extract()?;
    let abstract_model: bool = meta.getattr("abstract")?.extract()?;
    let proxy: bool = meta.getattr("proxy")?.extract()?;

    let parents = extract_parents(&meta)?;
    let fields = extract_fields(py, &meta)?;
    let relations = extract_relations(py, &meta)?;
    let managers = extract_managers(py, model_cls)?;
    let methods = extract_methods(py, model_cls)?;

    Ok(DjangoModel {
        app_label,
        name,
        db_table,
        module,
        source_file: String::new(),
        abstract_model,
        proxy,
        fields,
        relations,
        managers,
        parents,
        methods,
    })
}

fn extract_parents(meta: &Bound<'_, PyAny>) -> PyResult<Vec<String>> {
    let parents_dict = meta.getattr("parents")?;
    let mut parents = Vec::new();
    for parent in parents_dict.call_method0("keys")?.try_iter()? {
        let parent = parent?;
        let parent_name: String = parent.getattr("__name__")?.extract()?;
        parents.push(parent_name);
    }
    Ok(parents)
}

fn extract_fields(_py: Python<'_>, meta: &Bound<'_, PyAny>) -> PyResult<Vec<FieldDef>> {
    let mut fields = Vec::new();

    let concrete_fields = meta.getattr("local_concrete_fields")?;
    for field in concrete_fields.try_iter()? {
        let field = field?;
        let field_class = get_class_path(&field)?;

        // Skip relational fields — handled in extract_relations
        if is_relational(&field_class) {
            continue;
        }

        let name: String = field.getattr("name")?.extract()?;
        let column: String = field.getattr("column")?.extract()?;
        let nullable: bool = field.getattr("null")?.extract()?;
        let blank: bool = field.getattr("blank")?.extract()?;
        let primary_key: bool = field.getattr("primary_key")?.extract()?;
        let unique: bool = field.getattr("unique")?.extract()?;
        let db_index: bool = field.getattr("db_index")?.extract()?;

        let max_length = field
            .getattr("max_length")
            .ok()
            .and_then(|v| v.extract::<i64>().ok());

        let native_type = field_class_to_python_type(&field_class);

        let default = extract_default(_py, &field)?;
        let choices = extract_choices(&field)?;
        let validators = extract_field_validators(&field)?;

        fields.push(FieldDef {
            name,
            column,
            field_class,
            native_type,
            nullable,
            blank,
            default,
            max_length,
            choices,
            validators,
            primary_key,
            unique,
            db_index,
        });
    }

    Ok(fields)
}

fn extract_relations(_py: Python<'_>, meta: &Bound<'_, PyAny>) -> PyResult<Vec<RelationDef>> {
    let mut relations = Vec::new();

    let kwargs = PyDict::new(_py);
    kwargs.set_item("include_hidden", true)?;
    let all_fields = meta.call_method("get_fields", (), Some(&kwargs))?;

    for field in all_fields.try_iter()? {
        let field = field?;
        let field_class = get_class_path(&field)?;

        let kind = if field_class.contains("ForeignKey") {
            RelationKind::ForeignKey
        } else if field_class.contains("OneToOneField") {
            RelationKind::OneToOne
        } else if field_class.contains("ManyToManyField") {
            RelationKind::ManyToMany
        } else if field_class.contains("ManyToOneRel") {
            RelationKind::Reverse
        } else if field_class.contains("OneToOneRel") {
            RelationKind::ReverseOneToOne
        } else if field_class.contains("ManyToManyRel") {
            RelationKind::ManyToMany
        } else {
            continue;
        };

        let name: String = field
            .getattr("name")
            .or_else(|_| field.call_method0("get_accessor_name"))
            .and_then(|v| v.extract())
            .unwrap_or_default();

        let related_model = field.getattr("related_model").or_else(|_| {
            field
                .getattr("remote_field")
                .and_then(|rf| rf.getattr("model"))
        });

        let (to_model, to_model_app) = match related_model {
            Ok(rm) => {
                let to_model: String = rm.getattr("__name__")?.extract()?;
                let to_model_app: String = rm.getattr("_meta")?.getattr("app_label")?.extract()?;
                (to_model, to_model_app)
            }
            Err(_) => continue,
        };

        let related_name: String = field
            .getattr("related_name")
            .and_then(|v| v.extract())
            .unwrap_or_default();

        let related_query_name: String = field
            .call_method0("get_related_field")
            .and_then(|v| v.getattr("name"))
            .and_then(|v| v.extract())
            .unwrap_or_default();

        let on_delete: Option<String> = field
            .getattr("remote_field")
            .ok()
            .and_then(|rf| rf.getattr("on_delete").ok())
            .and_then(|od| od.getattr("__name__").ok())
            .and_then(|n| n.extract().ok());

        let nullable: bool = field
            .getattr("null")
            .and_then(|v| v.extract())
            .unwrap_or(false);

        let through_model: Option<String> = if kind == RelationKind::ManyToMany {
            field
                .getattr("remote_field")
                .ok()
                .and_then(|rf| rf.getattr("through").ok())
                .and_then(|t| t.getattr("__name__").ok())
                .and_then(|n| n.extract().ok())
        } else {
            None
        };

        relations.push(RelationDef {
            name,
            kind,
            to_model,
            to_model_app,
            related_name,
            related_query_name,
            on_delete,
            nullable,
            through_model,
        });
    }

    Ok(relations)
}

fn extract_managers(_py: Python<'_>, model_cls: &Bound<'_, PyAny>) -> PyResult<Vec<ManagerDef>> {
    let meta = model_cls.getattr("_meta")?;
    let managers = meta.getattr("managers")?;
    let default_manager = meta.getattr("default_manager")?;

    let mut result = Vec::new();

    for manager in managers.try_iter()? {
        let manager = manager?;
        let name: String = manager.getattr("name")?.extract()?;
        let is_default = manager.is(&default_manager);
        let manager_class = get_class_path(&manager)?;

        let queryset_class: String = manager
            .getattr("_queryset_class")
            .and_then(|qs| get_class_path(&qs))
            .unwrap_or_else(|_| "django.db.models.QuerySet".into());

        let custom_methods = extract_public_methods(_py, &manager)?;

        result.push(ManagerDef {
            name,
            manager_class,
            queryset_class,
            is_default,
            custom_methods,
        });
    }

    Ok(result)
}

fn extract_methods(py: Python<'_>, model_cls: &Bound<'_, PyAny>) -> PyResult<Vec<String>> {
    let mut methods = extract_public_methods(py, model_cls)?;
    // Also include @property descriptors — they act as virtual fields
    let builtins = py.import("builtins")?;
    let property_type = builtins.getattr("property")?;
    for cls in model_cls.getattr("__mro__")?.try_iter()? {
        let cls = cls?;
        let cls_name: String = cls.getattr("__name__")?.extract()?;
        if cls_name == "Model" || cls_name == "object" {
            break;
        }
        if let Ok(dict) = cls.getattr("__dict__") {
            if let Ok(items) = dict.call_method0("items") {
                for item in (items.try_iter()?).flatten() {
                    if let Ok((key, val)) = item.extract::<(String, pyo3::Bound<'_, pyo3::PyAny>)>()
                    {
                        if !key.starts_with('_')
                            && val.is_instance(&property_type).unwrap_or(false)
                            && !methods.contains(&key)
                        {
                            methods.push(key);
                        }
                    }
                }
            }
        }
    }
    Ok(methods)
}

fn extract_installed_apps(py: Python<'_>) -> PyResult<Vec<String>> {
    let settings = py.import("django.conf")?.getattr("settings")?;
    let apps: Vec<String> = settings.getattr("INSTALLED_APPS")?.extract()?;
    Ok(apps)
}

fn extract_settings(py: Python<'_>) -> PyResult<DjangoSettings> {
    let settings = py.import("django.conf")?.getattr("settings")?;

    let auth_user_model: String = settings
        .getattr("AUTH_USER_MODEL")
        .and_then(|v| v.extract())
        .unwrap_or_else(|_| "auth.User".into());

    let databases: Vec<String> = settings
        .getattr("DATABASES")
        .and_then(|d| d.call_method0("keys"))
        .and_then(|k| {
            let mut keys = Vec::new();
            for item in k.try_iter()? {
                let item = item?;
                keys.push(item.extract::<String>()?);
            }
            Ok(keys)
        })
        .unwrap_or_default();

    let middleware: Vec<String> = settings
        .getattr("MIDDLEWARE")
        .and_then(|v| v.extract())
        .unwrap_or_default();

    // Store framework-specific settings in the extra map
    let mut extra = std::collections::HashMap::new();
    if let Ok(drf) = settings.getattr("REST_FRAMEWORK") {
        if let Ok(auth) = drf
            .get_item("DEFAULT_AUTHENTICATION_CLASSES")
            .and_then(|v| v.str())
        {
            extra.insert(
                "drf.default_authentication_classes".into(),
                auth.to_string(),
            );
        }
        if let Ok(perm) = drf
            .get_item("DEFAULT_PERMISSION_CLASSES")
            .and_then(|v| v.str())
        {
            extra.insert("drf.default_permission_classes".into(), perm.to_string());
        }
        if let Ok(page) = drf
            .get_item("DEFAULT_PAGINATION_CLASS")
            .and_then(|v| v.extract::<String>())
        {
            extra.insert("drf.default_pagination_class".into(), page);
        }
    }

    Ok(DjangoSettings {
        auth_user_model,
        databases,
        middleware,
        extra,
    })
}

// ── Helpers ───────────────────────────────────────────────────────────────

fn get_class_path(obj: &Bound<'_, PyAny>) -> PyResult<String> {
    let cls = obj.getattr("__class__")?;
    let module: String = cls.getattr("__module__")?.extract()?;
    let name: String = cls.getattr("__name__")?.extract()?;
    Ok(format!("{module}.{name}"))
}

fn is_relational(field_class: &str) -> bool {
    field_class.contains("ForeignKey")
        || field_class.contains("OneToOneField")
        || field_class.contains("ManyToManyField")
        || field_class.contains("Rel")
}

fn field_class_to_python_type(field_class: &str) -> String {
    if field_class.contains("IntegerField") || field_class.contains("AutoField") {
        "int".into()
    } else if field_class.contains("FloatField") || field_class.contains("DecimalField") {
        "float".into()
    } else if field_class.contains("BooleanField") {
        "bool".into()
    } else if field_class.contains("DateTimeField") {
        "datetime".into()
    } else if field_class.contains("DateField") {
        "date".into()
    } else if field_class.contains("TimeField") {
        "time".into()
    } else if field_class.contains("UUIDField") {
        "UUID".into()
    } else if field_class.contains("JSONField") {
        "Any".into()
    } else if field_class.contains("BinaryField") {
        "bytes".into()
    } else {
        "str".into()
    }
}

fn extract_default(py: Python<'_>, field: &Bound<'_, PyAny>) -> PyResult<Option<String>> {
    let django_fields = py.import("django.db.models.fields")?;
    let not_provided = django_fields.getattr("NOT_PROVIDED")?;
    let default = field.getattr("default")?;

    if default.is(&not_provided) {
        return Ok(None);
    }

    if default.is_callable() {
        let name: String = default
            .getattr("__name__")
            .or_else(|_| default.getattr("__class__")?.getattr("__name__"))
            .and_then(|v| v.extract())
            .unwrap_or_else(|_| "unknown".into());
        return Ok(Some(format!("<callable: {name}>")));
    }

    Ok(Some(format!("{default}")))
}

fn extract_choices(field: &Bound<'_, PyAny>) -> PyResult<Vec<(String, String)>> {
    // Prefer flatchoices which is always a flat list of (value, label) pairs,
    // even for grouped choices and TextChoices/IntegerChoices enums.
    let choices_obj = field
        .getattr("flatchoices")
        .or_else(|_| field.getattr("choices"))?;

    if choices_obj.is_none() {
        // choices is Python None — no choices defined
        return Ok(vec![]);
    }

    // For EnumChoiceField the choices may live on an enum_class attribute.
    // If the resolved object is not iterable or is empty, fall back to that.
    let mut result = collect_flat_choices(&choices_obj);

    if result.is_empty() {
        // Try EnumChoiceField: field.enum_class.choices
        if let Ok(enum_cls) = field.getattr("enum_class") {
            if !enum_cls.is_none() {
                if let Ok(enum_choices) = enum_cls.getattr("choices") {
                    result = collect_flat_choices(&enum_choices);
                }
            }
        }
    }

    Ok(result)
}

/// Collect (value, label) pairs from a Django choices iterable.
/// Handles both flat choices `[(val, label), ...]` and grouped choices
/// `[(group, [(val, label), ...]), ...]`.
fn collect_flat_choices(choices_obj: &Bound<'_, PyAny>) -> Vec<(String, String)> {
    let mut result = Vec::new();
    let Ok(iter) = choices_obj.try_iter() else {
        return result;
    };
    for item in iter.flatten() {
        let Ok(label_or_group) = item.get_item(1) else {
            continue;
        };
        // Grouped choices: item[1] is a list/tuple of (value, label) pairs
        if label_or_group.try_iter().is_ok()
            && !label_or_group.is_instance_of::<pyo3::types::PyString>()
        {
            // This is a group — recurse into sub-choices
            for sub in label_or_group.try_iter().into_iter().flatten().flatten() {
                let value = sub.get_item(0).map(|v| format!("{v}")).unwrap_or_default();
                let label = sub.get_item(1).map(|v| format!("{v}")).unwrap_or_default();
                if !value.is_empty() || !label.is_empty() {
                    result.push((value, label));
                }
            }
        } else {
            // Flat choice: item = (value, label)
            let value = item.get_item(0).map(|v| format!("{v}")).unwrap_or_default();
            let label = format!("{label_or_group}");
            if !value.is_empty() || !label.is_empty() {
                result.push((value, label));
            }
        }
    }
    result
}

fn extract_field_validators(field: &Bound<'_, PyAny>) -> PyResult<Vec<String>> {
    let validators = field.getattr("validators")?;
    let mut result = Vec::new();
    for v in validators.try_iter()? {
        let v = v?;
        result.push(get_class_path(&v)?);
    }
    Ok(result)
}

fn extract_public_methods(_py: Python<'_>, obj: &Bound<'_, PyAny>) -> PyResult<Vec<String>> {
    let dir = obj.dir()?;
    let mut methods = Vec::new();
    for item in dir.iter() {
        let name_str: String = item.extract()?;
        // Skip private methods (_foo) but keep dunder methods (__str__, __repr__, etc.)
        if name_str.starts_with('_') && !name_str.starts_with("__") {
            continue;
        }
        if let Ok(attr) = obj.getattr(name_str.as_str()) {
            if attr.is_callable() {
                methods.push(name_str);
            }
        }
    }
    Ok(methods)
}

fn capitalize(s: &str) -> String {
    let mut c = s.chars();
    match c.next() {
        None => String::new(),
        Some(f) => f.to_uppercase().collect::<String>() + c.as_str(),
    }
}
